### PR TITLE
Support more compliation targets (including WebAssembly, `no_std`, and `miri`)

### PR DESCRIPTION
**Pull request by robamler**
I'm curious if you'd be open to considering a pair of related pull requests to both `special` and `probability` that make both libraries more portable by adding some cargo "features". I think the functionality provided by both `special` and `probability` is so fundamental that both crates would be very useful on many platforms (e.g., I currently have two students who're working on a project that will eventually use `probability` from WebAssembly).

The downside of supporting these additional compilation targets is that it will require an extra round of testing for each future release (in addition to running `cargo test`, one would also have to run `cargo test --no-default-features`).

This PR is split into two commits that add declarations for features `system_math` and `std`, respectively, to `Cargo.toml`. Both features are activated by default to make the changes both backward and forward compatible, but they can be deactivated by downstream crates:

* Deactivating the feature `system_math` replaces `special`'s dependency on the system "libm" C library with a re-implementation in pure rust. This allows `special` to be used on compilation targets that don't have a system "libm" C library, such as WebAssembly.
* Deactivating the feature `std` allows applications that use `special` to be compiled in `no_std` mode (e.g., for microcontrollers).

A positive side effect of these changes is that they make it possible to run `miri` on this crate and potentially on downstream crates (`miri` is a tool to detect undefined behavior in unsafe code): run `cargo +nightly miri test --no-default-features`

The companion PR for `probability` is at stainless-steel/probability#27.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,8 +18,19 @@ readme = "README.md"
...
+# use the system math C library for mathematical operations. Deactivate this
+# default feature if no system math C library is available on your compilation
+# target (e.g., if you compile to WebAssembly, or if you want test for undefined
+# behaviour using `miri`).
+system_math = ["libc"]
+
 [dependencies]
-libc = "0.2"
+libc = {version = "0.2", optional = true} 
+libm = "0.2.2" # Only used if default feature "libc" is *not* activated.
```
Comment (robamler) [inline Cargo.toml:38]: Should actually read: `# Only used if default feature "system_math" is *not* activated.`

**Comment by IvanUkhov** [comment]
I apologize for the late response. It was too much to process. Hence, the procrastination. I have a few questions.

If there is a re-implementation in pure Rust, why keeping the dependency on `libc` at all? Is it because of totally different performance profiles?

If `no_std` allows one to implement the current function set, why not getting rid of `std` all together?

**Comment by robamler** [comment]
Sorry for the late response on my side.

I'd be happy to remove the feature gates if you'd prefer that. My motivation for these gates was not so much driven by performance but more by backward and forward compatibility. The following considerations were on my mind:

- **backward compatibility**:
  - *(my main motivation for introducing feature gates:)* Changing the underlying implementation of special mathematical functions will likely lead to slightly different results of calculations since different implementations might use varying approximations. This will break any downstream crates that require *exactly* reproducible calculations. For example, I maintain the [`constriction` crate](https://crates.io/crates/constriction), for which this would be an issue. A more common place where even tiny approximation errors might become an issue could be unit tests in downstream crates if they check results of example calculations in a naive way (those unit tests could be fixed, though).
  - Since the proposed changes have a fairly "global" effect I can't say with certainty that *all* functionality really still works with both default features turned off. All unit tests pass with the two default features turned off, but there might be untested functionality that breaks (it's unlikely though since it's hard to think of an error in this context that wouldn't already show up at compilation time).
- **forward compatibility:**
  - Even though all current functionality still seems to work with the two default features turned off, some new feature that you might want to add in a future version of `special` or `probability` might not. Removing the proposed feature gates would therefore lock you in to a less capable programming environment. By contrast, keeping the proposed feature gates with the proposed defaults would make sure that nothing changes in the "default" case. This would allow you to add new functionality even if it doesn't work in `no_std` mode or if it requires a system `libm` library. Such new functionality then couldn't be used by the few downstream crates that turn off the `std` or `system_math` feature, but most users will be able to get the new functionality.

Of course, removing the feature gates would have the advantage of not splitting the ecosystem, so I'd be happy either way.

### Concerning Performance

I added a commit with benchmarking code for four affected functions. Below are my results when I run the benchmark 100 times with default settings (blue) and 100 times with the `--no-default-features` switch (orange); these 200 runs were interleaved randomly.

benches

It's not a clear picture: the system math library is faster (as one would expect) for 3 out of 4 functions, but the situation is reversed for `ln_gamma()`. From a backward-compatibility standpoint, silently switching out the system `libm` for its reimplementation in Rust would lead to a performance regression, at least for some code.

(The above benchmark results might depend strongly on the operating system that provides the system math library; above results were obtained on Ubuntu 22.04.1 with libc version 2.35 and Rust version 1.65.0-nightly (ce40690a5 2022-08-09)).

**Comment by IvanUkhov** [comment]
Thank you for the answers! I would prefer to have one implementation, and it should be a Rust one whenever possible. As for the backward and forward compatibilities, I would not be afraid to break things at this stage. It is better to have clean code in a state we want it to be in instead of maintaining a bag of special cases.

If you have time, please go ahead and reduce it to `libm` with `no_std`. Let us please also not extend the functionality of this crate within this PR. (I am referring to `FloatExt`.) It can be discussed separately.

**Comment by IvanUkhov** [comment]
I have now updated the crate with respect to my previous comment. Thank you for your PR!

**Pull request merged**